src/trace-server.ts

```diff
@@ -168,27 +168,57 @@ export class TraceServer {
             },
             async progress => {
                 progress.report({ message: 'starting up...' });
-                let timeout = false;
-                const timeoutId = setTimeout(() => (timeout = true), millis);
-
-                // eslint-disable-next-line no-constant-condition
-                while (true) {
-                    if (await this.isUp()) {
-                        this.showStatus(true);
-                        clearTimeout(timeoutId);
-                        break;
-                    }
-                    if (timeout) {
-                        this.showError(prefix + ' startup timed-out after ' + millis + 'ms.');
-                        await this.stopOrReset(context);
-                        break;
+                try {
+                    let timeout = false;
+                    const timeoutId = setTimeout(() => (timeout = true), millis);
+
+                    let stderr = '';
+                    this.server?.stderr?.on('data', data => {
+                        stderr = data;
+                    });
+
+                    let errorString: string | undefined;
+                    this.server?.once('exit', code => {
+                        errorString = 'Code: ' + code + (stderr !== '' ? '\nstdErr: ' + stderr : '');
+                    });
+
+                    this.server?.once('error', () => {
+                        errorString = 'Error occurred.' + (stderr !== '' ? '\nstdErr: ' + stderr : '');
+                    });
+
+                    // eslint-disable-next-line no-constant-condition
+                    while (true) {
+                        if (await this.isUp()) {
+                            clearTimeout(timeoutId);
+                            this.showStatus(true);
+                            break;
+                        }
+
+                        // Check if trace server exited with error
+                        if (errorString !== undefined) {
+                            throw new Error(errorString);
+                        }
+
+                        // Check for overall timeout
+                        if (timeout) {
+                            throw new Error(prefix + ' startup timed-out after ' + millis + 'ms.');
+                        }
+
+                        // Wait before trying again
+                        await new Promise<void>(resolve => setTimeout(() => resolve(), 1000));
                     }
+                } catch (error) {
+                    this.showError(prefix + ' startup failed: ' + error);
+                    await this.stopOrReset(context);
+                    return;
+                } finally {
+                    this.server?.removeAllListeners();
```

Comment:
From what I read, this is not the preferred way to unregistered listeners, since, depending on timing, it might unregistered listeners registered elsewhere and have unforeseen side-effects. 

Instead we might use "this.server?.off()" to de-register listener(s) here?